Create IstioRevisionTag documentation (#511)
istio-ecosystem/sail-operator: docs/README.md

In Istio, stable revision tags are usually created using `istioctl`, but if you're using the Sail Operator, you can use the `IstioRevisionTag` resource, which comes with an additional feature: instead of just being able to reference an `IstioRevision`, you can also reference an `Istio` resource. When you now update your control plane and the underlying `IstioRevision` changes, the Sail Operator will update your revision tag for you. You only need to restart your deployments to re-inject the new proxies.

```yaml
apiVersion: sailoperator.io/v1alpha1
kind: IstioRevisionTag
metadata:
  name: default
spec:
  targetRef:
    kind: Istio   # can be either Istio or IstioRevision
    name: prod    # the name of the Istio/IstioRevision resource
```

> As you can see in the YAML above, `IstioRevisionTag` really only has one field in its spec: `targetRef`. With this field, you can reference an `Istio` or `IstioRevision` resource. So after deploying this, you will be able to use both the `istio.io/rev=default` and also `istio-injection=enabled` labels to inject proxies into your workloads. The `istio-injection` label can only be used for revisions named `default` - and this tag is called `default`.

Reviewer: "for revisions named `default` - and this tag is called `default`"
It might be a little bit confusing. 
Maybe something like "can only be used for revisions named `default` or when you create a IstioRevisionTag named `default.`"

Author: Fixed!

      - kube-system
```

### Resource Status
All of the Sail Operator API resources have a `status` subresource that contains information about their current state in the Kubernetes cluster.

#### Conditions
All resources have a `Ready` condition which is set to `true` as soon as all child resource have been created and are deemed Ready by their respective controllers. To see additional conditions for each of the resources, check the [API reference documentation](https://github.com/istio-ecosystem/sail-operator/tree/main/docs/api-reference/sailoperator.io.md).

> #### InUse Detection

Reviewer: Could you please add a few sentences when this is useful from user point of view?

Reviewer: Ah, ok, it's used in the example below. So maybe just one sentence: "See the example bellow for proper usage."

10. Verify the `Istio` and `IstioRevision` resources. There will be a new revision created with the new version.

    ```console

>     $ kubectl get istio

Reviewer: Why removing the namespace here?

Author: istio resource is cluster-scoped

Reviewer: Ah right. Sorry.

    NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   STATUS    VERSION   AGE
    default   1           1       1        default-v1-23-2   Healthy   v1.23.2   5m

    $ kubectl get istiorevision
    NAME              TYPE    READY   STATUS    IN USE   VERSION   AGE
    default-v1-23-2   Local   True    Healthy   True     v1.23.2   5m31s
    ```
    The old `IstioRevision` resource and the old control plane will be deleted when the grace period specified in the `Istio` resource field `spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds` expires.

#### Example using the RevisionBased strategy and an IstioRevisionTag

Prerequisites:
* Sail Operator is installed.
* `istioctl` is [installed](common/install-istioctl-tool.md).

Steps:

1. Create the `istio-system` namespace.

    ```bash
    kubectl create namespace istio-system
    ```

2. Create the `Istio` and `IstioRevisionTag` resources.

    ```bash
    cat <<EOF | kubectl apply -f-
    apiVersion: sailoperator.io/v1alpha1
    kind: Istio
    metadata:
      name: default
    spec:
      namespace: istio-system
      updateStrategy:
        type: RevisionBased
        inactiveRevisionDeletionGracePeriodSeconds: 30

>       version: v1.22.5

Reviewer: ```suggestion
      version: v1.23.3
```

Author: Changed!

    ---
    apiVersion: sailoperator.io/v1alpha1
    kind: IstioRevisionTag
    metadata:
      name: default
    spec:
      targetRef:
        kind: Istio
        name: default
    EOF
    ```

3. Confirm the control plane is installed and is using the desired version.

    ```console
    $ kubectl get istio
    NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   STATUS    VERSION   AGE

>     default   1           1       1        default-v1-22-5   Healthy   v1.22.5   52s

Reviewer: ```suggestion
    default   1           1       1        default-v1-22-5   Healthy   v1.23.3   52s
```

Author: Changed!

    ```
    Note: `IN USE` field shows as 1, even though no workloads are using the control plane. This is because the `IstioRevisionTag` is referencing it.

4. Inspect the `IstioRevisionTag`.

    ```console
    $ kubectl get istiorevisiontags
    NAME      STATUS                    IN USE   REVISION          AGE

>     default   NotReferencedByAnything   False    default-v1-22-5   52s

Author: Changed!

    ```

5. Create `bookinfo` namespace and label it to mark it for injection.

    ```bash
    kubectl create namespace bookinfo
    kubectl label namespace bookinfo istio-injection=enabled
    ```

6. Deploy bookinfo application.

    ```bash

>     kubectl apply -n bookinfo -f https://raw.githubusercontent.com/istio/istio/release-1.22/samples/bookinfo/platform/kube/bookinfo.yaml

Reviewer: ```suggestion
    kubectl apply -n bookinfo -f https://raw.githubusercontent.com/istio/istio/release-1.23/samples/bookinfo/platform/kube/bookinfo.yaml
```

Author: Changed!

    ```

7. Review the `IstioRevisionTag` resource after application deployment.

    ```console
    $ kubectl get istiorevisiontag
    NAME      STATUS    IN USE   REVISION          AGE

>     default   Healthy   True     default-v1-22-5   2m46s

Reviewer: ```suggestion
    default   Healthy   True     default-v1-23-3   2m46s
```

Author: Changed!

    ```
    Note: `IN USE` field shows 'True', as the tag is now referenced by both active workloads and the bookinfo namespace.

8. Confirm that the proxy version matches the control plane version.

    ```bash
    istioctl proxy-status 
    ```
    The column `VERSION` should match the control plane version.

9. Update the control plane to a new version.

    ```bash

>     kubectl patch istio default -n istio-system --type='merge' -p '{"spec":{"version":"v1.23.2"}}'

Reviewer: ```suggestion
    kubectl patch istio default -n istio-system --type='merge' -p '{"spec":{"version":"v1.24.1"}}'
```

Author: Changed!

    ```

10. Verify the `Istio`, `IstioRevision` and `IstioRevisionTag` resources. There will be a new revision created with the new version.

    ```console
    $ kubectl get istio
    NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   STATUS    VERSION   AGE

>     default   2           2       1        default-v1-23-2   Healthy   v1.23.2   9m23s

Author: Changed!

    $ kubectl get istiorevision
    NAME              TYPE    READY   STATUS    IN USE   VERSION   AGE

>     default-v1-22-5   Local   True    Healthy   True     v1.22.5   10m

Author: Changed!

>     default-v1-23-2   Local   True    Healthy   True    v1.23.2   66s

Reviewer: ```suggestion
    default-v1-24-1   Local   True    Healthy   True    v1.24.1   66s
```

Author: Changed!

    $ kubectl get istiorevisiontag
    NAME      STATUS    IN USE   REVISION          AGE

>     default   Healthy   True     default-v1-23-2   10m44s

Reviewer: ```suggestion
    default   Healthy   True     default-v1-24-1   10m44s
```

Author: Changed!

    ```

>     Now, both our IstioRevisions and the IstioRevisionTag are considered in use. The old revision default-v1-22-5 because it is being used by proxies, the new revision default-v1-23-2 because it is referenced by the tag, and lastly the tag because it is referenced by the bookinfo namespace.

Reviewer: ```suggestion
    Now, both our IstioRevisions and the IstioRevisionTag are considered in use. The old revision default-v1-23-3 because it is being used by proxies, the new revision default-v1-24-1 because it is referenced by the tag, and lastly the tag because it is referenced by the bookinfo namespace.
```

Author: Changed!

11. Confirm there are two control plane pods running, one for each revision.

    ```console
    $ kubectl get pods -n istio-system
    NAME                                      READY   STATUS    RESTARTS   AGE

>     istiod-default-v1-22-5-c98fd9675-r7bfw    1/1     Running   0          10m

Reviewer: ```suggestion
    istiod-default-v1-23-3-c98fd9675-r7bfw    1/1     Running   0          10m
```

>     istiod-default-v1-23-2-7495cdc7bf-v8t4g   1/1     Running   0          113s

Reviewer: ```suggestion
    istiod-default-v1-24-1-7495cdc7bf-v8t4g   1/1     Running   0          113s
```

Author: Changed!

    $ kubectl get istiorevisiontag
    NAME      STATUS    IN USE   REVISION          AGE
    default   Healthy   True     default-v1-24-1   10m44s
    ```
    Now, both our IstioRevisions and the IstioRevisionTag are considered in use. The old revision default-v1-23-3 because it is being used by proxies, the new revision default-v1-24-1 because it is referenced by the tag, and lastly the tag because it is referenced by the bookinfo namespace.

11. Confirm there are two control plane pods running, one for each revision.

    ```console
    $ kubectl get pods -n istio-system
    NAME                                      READY   STATUS    RESTARTS   AGE
    istiod-default-v1-23-3-c98fd9675-r7bfw    1/1     Running   0          10m
    istiod-default-v1-24-1-7495cdc7bf-v8t4g   1/1     Running   0          113s
    ```

12. Confirm the proxy sidecar version remains the same:

    ```bash
    istioctl proxy-status 
    ```
    The column `VERSION` should still match the old control plane version.

13. Restart all the Deployments in the `bookinfo` namespace.

    ```bash
    kubectl rollout restart deployment -n bookinfo
    ```

> 14. Confirm the new version is used in the sidecars.

Reviewer: ```suggestion
14. Confirm the new version is used in the sidecars. Wait a few seconds, it might take a while until the bookinfo pods are restarted.
```
